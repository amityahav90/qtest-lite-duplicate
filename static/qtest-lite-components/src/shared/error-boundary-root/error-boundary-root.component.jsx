import React from 'react';
import PropTypes from 'prop-types';
import { ErrorBoundary } from '../error-boundary';

export class ErrorBoundaryRoot extends ErrorBoundary {
	constructor(props) {
		super(props);

		this.boundWindowErrorFn = this.handleWindowError.bind(this);
		this.boundWindowUEFn = this.handleWindowUnhandledException.bind(this);
	}

	componentDidMount() {
		window.addEventListener('error', this.boundWindowErrorFn);
		window.addEventListener('unhandledrejection', this.boundWindowUEFn);
	}

	componentWillUnmount() {
		window.removeEventListener('error', this.boundWindowErrorFn);
		window.removeEventListener('unhandledrejection', this.boundWindowUEFn);
	}

	handleWindowError(e) {
		const error = e ? (e.error ? e.error.stack : e.error) : e;

		this.log.error('Unhandled exception', error);

		// Prevent log error to console
		e.preventDefault();
	}

	handleWindowUnhandledException(e) {
		const error = e ? (e.reason ? e.reason.stack : e.reason) : e;

		this.log.error('Unhandled promise rejection', error);

		// Prevent log error to console
		e.preventDefault();
	}
}

ErrorBoundaryRoot.propTypes = {
	children: PropTypes.any.isRequired,
	logPath: PropTypes.string.isRequired,
	fallbackRender: PropTypes.func
};
