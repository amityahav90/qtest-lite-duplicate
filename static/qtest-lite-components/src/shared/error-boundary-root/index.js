export * from './error-boundary-root.component';
export * from './with-error-boundary-root';
