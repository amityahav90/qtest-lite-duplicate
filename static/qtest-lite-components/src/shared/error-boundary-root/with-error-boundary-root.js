import React from 'react';
import { LogContext } from '../log-service';
import { ErrorBoundaryRoot } from './error-boundary-root.component';

export function withErrorBoundaryRoot(WrappedComponent, fallbackRender) {
	function wrapped(props) {
		return (
			<LogContext.Consumer>
				{(value) => (
					<ErrorBoundaryRoot logPath={value} fallbackRender={fallbackRender}>
						<WrappedComponent {...props} />
					</ErrorBoundaryRoot>
				)}
			</LogContext.Consumer>
		);
	}

	return wrapped;
}
