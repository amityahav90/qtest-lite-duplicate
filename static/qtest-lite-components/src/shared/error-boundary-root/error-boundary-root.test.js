import React from 'react';
import renderer from 'react-test-renderer';
import { ErrorBoundaryRoot } from './error-boundary-root.component';

describe('Test ErrorBoundaryRoot component', () => {
	beforeAll(() => {
		jest.mock('@forge/bridge');
	});

	test('Component rendered correctly', async () => {
		const component = renderer.create(
			<ErrorBoundaryRoot logPath="test">
				<div>test</div>
			</ErrorBoundaryRoot>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});

	test('"fallbackRender" functionality rendered correctly', async () => {
		function ThrowError() {
			throw new Error('fallbackRender functionality');
		}

		function FallbackRender() {
			return <div>fallbackRender functionality worked</div>;
		}

		const component = renderer.create(
			<ErrorBoundaryRoot logPath="test" fallbackRender={() => <FallbackRender />}>
				<ThrowError />
			</ErrorBoundaryRoot>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});
});
