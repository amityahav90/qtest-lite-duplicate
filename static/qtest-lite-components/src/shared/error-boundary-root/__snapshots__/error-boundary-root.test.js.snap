// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ErrorBoundaryRoot component "fallbackRender" functionality rendered correctly 1`] = `
<div>
  fallbackRender functionality worked
</div>
`;

exports[`Test ErrorBoundaryRoot component Component rendered correctly 1`] = `
<div>
  test
</div>
`;
